File: packages/plugin-agent/src/server/tools/laps.ts
import type { AgentToolDefinition, AgentToolHandler, AgentToolContext, AgentToolResult } from '@purplesector/plugin-api';

export const lapToolDefinitions: AgentToolDefinition[] = [
  {
    name: 'listLaps',
    description: 'List all laps for a session, ordered by lap number.',
    inputSchema: {
      type: 'object',
      properties: {
        sessionId: { type: 'string', description: 'The session ID' },
      },
      required: ['sessionId'],
    },
    category: 'laps',
    mutating: false,
  },
  {
    name: 'getLap',
    description: 'Get details of a specific lap by ID (without full telemetry data).',
    inputSchema: {
      type: 'object',
      properties: {
        lapId: { type: 'string', description: 'The lap ID' },
      },
      required: ['lapId'],
    },
    category: 'laps',
    mutating: false,
  },
  {
    name: 'getLapTelemetry',
    description: 'Get the full telemetry data for a specific lap.',
    inputSchema: {
      type: 'object',
      properties: {
        lapId: { type: 'string', description: 'The lap ID' },
      },
      required: ['lapId'],
    },
    category: 'laps',
    mutating: false,
  },
];

export function createLapToolHandlers(prisma: any): Record<string, AgentToolHandler> {
  return {
    async listLaps(args: Record<string, unknown>, ctx: AgentToolContext): Promise<AgentToolResult> {
      const session = await prisma.session.findFirst({
        where: { id: args.sessionId as string, userId: ctx.userId },
      });
      if (!session) return { success: false, message: 'Session not found.' };

      const laps = await prisma.lap.findMany({
        where: { sessionId: args.sessionId as string },
        select: {
          id: true,
          lapNumber: true,
          lapTime: true,
          analyzed: true,
          tags: true,
          createdAt: true,
        },
        orderBy: { lapNumber: 'asc' },
      });
      return { success: true, data: laps, message: `Found ${laps.length} laps.` };
    },

    async getLap(args: Record<string, unknown>, ctx: AgentToolContext): Promise<AgentToolResult> {
      const lap = await prisma.lap.findFirst({
        where: { id: args.lapId as string, userId: ctx.userId },
        select: {
          id: true,
          lapNumber: true,
          lapTime: true,
          analyzed: true,
          suggestions: true,
          driverComments: true,
          tags: true,
          createdAt: true,
          session: { select: { id: true, name: true, eventId: true } },
        },
      });
      if (!lap) return { success: false, message: 'Lap not found.' };
      return { success: true, data: lap };
    },

    async getLapTelemetry(args: Record<string, unknown>, ctx: AgentToolContext): Promise<AgentToolResult> {
      const lap = await prisma.lap.findFirst({
        where: { id: args.lapId as string, userId: ctx.userId },
        select: { id: true, lapNumber: true, lapTime: true, telemetryData: true },
      });
      if (!lap) return { success: false, message: 'Lap not found.' };

      let telemetry;
      try {
        telemetry = JSON.parse(lap.telemetryData);
      } catch {
        telemetry = [];
      }
      return {
        success: true,
        data: { ...lap, telemetryData: undefined, telemetry, frameCount: telemetry.length },
        message: `Lap ${lap.lapNumber} has ${telemetry.length} telemetry frames.`,
      };
    },
  };
}

File: apps/web/apps/agent/tools/laps.py
"""Agent tools for Lap queries."""

from __future__ import annotations

from apps.telemetry.models import Lap
from apps.events.models import Session

from . import ToolDefinition, ToolContext, ToolResult, register_tools

definitions = [
    ToolDefinition(
        name="listLaps",
        description="List all laps for a session, ordered by lap number.",
        input_schema={
            "type": "object",
            "properties": {
                "sessionId": {"type": "string", "description": "The session ID"},
            },
            "required": ["sessionId"],
        },
        category="laps",
    ),
    ToolDefinition(
        name="getLap",
        description="Get details of a specific lap by ID (without full telemetry data).",
        input_schema={
            "type": "object",
            "properties": {
                "lapId": {"type": "string", "description": "The lap ID"},
            },
            "required": ["lapId"],
        },
        category="laps",
    ),
    ToolDefinition(
        name="getLapTelemetry",
        description="Get the full telemetry data for a specific lap.",
        input_schema={
            "type": "object",
            "properties": {
                "lapId": {"type": "string", "description": "The lap ID"},
            },
            "required": ["lapId"],
        },
        category="laps",
    ),
]


async def list_laps(args: dict, ctx: ToolContext) -> ToolResult:
    if not Session.objects.filter(id=args["sessionId"], user_id=ctx.user_id).exists():
        return ToolResult(success=False, message="Session not found.")
    laps = Lap.objects.filter(session_id=args["sessionId"]).order_by("lap_number")
    data = [
        {
            "id": str(lap.id),
            "lapNumber": lap.lap_number,
            "lapTime": lap.lap_time,
            "analyzed": lap.analyzed,
            "tags": lap.tags,
            "createdAt": lap.created_at.isoformat(),
        }
        for lap in laps
    ]
    return ToolResult(success=True, data=data, message=f"Found {len(data)} laps.")


async def get_lap(args: dict, ctx: ToolContext) -> ToolResult:
    try:
        lap = Lap.objects.select_related("session").get(id=args["lapId"])
    except Lap.DoesNotExist:
        return ToolResult(success=False, message="Lap not found.")
    return ToolResult(
        success=True,
        data={
            "id": str(lap.id),
            "lapNumber": lap.lap_number,
            "lapTime": lap.lap_time,
            "analyzed": lap.analyzed,
            "suggestions": lap.suggestions,
            "driverComments": lap.driver_comments,
            "tags": lap.tags,
            "createdAt": lap.created_at.isoformat(),
            "session": {
                "id": str(lap.session.id),
                "name": lap.session.name,
                "eventId": str(lap.session.event_id),
            },
        },
    )


async def get_lap_telemetry(args: dict, ctx: ToolContext) -> ToolResult:
    if not Lap.objects.filter(id=args["lapId"]).exists():
        return ToolResult(success=False, message="Lap not found.")
    return ToolResult(
        success=False,
        message="Lap telemetry frames are stored in Iceberg and are not available via this tool yet.",
    )


handlers = {
    "listLaps": list_laps,
    "getLap": get_lap,
    "getLapTelemetry": get_lap_telemetry,
}

register_tools(definitions, handlers)
